horovod/common/operations.cc

```diff
@@ -1709,15 +1459,52 @@ bool RunLoopOnce(HorovodGlobalState& state, MPIContext& ctx,
 void InitializeHorovodOnce(const int* ranks, int nranks) {
   // Ensure background thread is only started once.
   if (!horovod_global.initialize_flag.test_and_set()) {
-    for (int i = 0; i < nranks; ++i) {
-      horovod_global.ranks.push_back(ranks[i]);
+    horovod_global.controller = std::shared_ptr<Controller>(new MPIController());
+    horovod_global.controller->SetRank(ranks, nranks);
+
+    // Reset initialization flag
+    horovod_global.initialization_done = false;
+
+    horovod_global.background_thread = std::thread(
+        BackgroundThreadLoop, std::ref(horovod_global));
+  }
+
+  // Wait to ensure that the background thread has finished initializing MPI.
+  while (!horovod_global.initialization_done) {
+    std::this_thread::sleep_for(std::chrono::milliseconds(1));
+  }
+
+  LOG(DEBUG) << "Background thread init done";
+}
+
+// Start Horovod background thread. Ensure that this is
+// only done once no matter how many times this function is called.
+void InitializeHorovodOnceWithMPIComm(const int* ranks, int nranks, MPI_Comm
+comm) {
+  // Ensure background thread is only started once.
+  if (!horovod_global.initialize_flag.test_and_set()) {
+    LOG(INFO) << "Init with mpi communicator";
+    horovod_global.controller = std::shared_ptr<Controller>(new MPIController());
+    horovod_global.controller->SetRank(ranks, nranks);
+
+    if (horovod_global.controller->GetControllerType() ==
```

Comment:
This cast can be avoid by adding the MPIComm as a constructor parameter to MPIController.

Reply:
Yeah, fixed.